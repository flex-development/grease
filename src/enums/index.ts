/**
 * @file Entry Point - Enums
 * @module grease/enums
 */

export { default as LogLevel } from './log-level'
export { default as ReleaseType } from './release-type'
