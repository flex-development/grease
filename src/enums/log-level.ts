/**
 * @file Enums - LogLevel
 * @module grease/enums/LogLevel
 */

import type * as consola from 'consola'

/**
 * Log levels.
 *
 * @see https://github.com/unjs/consola#log-level
 *
 * @enum {consola.LogLevel}
 */
const LogLevel = Object.freeze({
  DEBUG: 4,
  ERROR: 0,
  FATAL: 0,
  INFO: 3,
  LOG: 2,
  SILENT: Number.NEGATIVE_INFINITY,
  TRACE: 5,
  VERBOSE: Number.POSITIVE_INFINITY,
  WARN: 1
} as const)

export default LogLevel
