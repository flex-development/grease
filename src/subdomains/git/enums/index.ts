/**
 * @file Entry Point - Enums
 * @module grease/git/enums
 */

export { default as RepositoryProvider } from './repository-provider'
