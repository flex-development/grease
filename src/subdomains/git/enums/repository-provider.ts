/**
 * @file Enums - RepositoryProvider
 * @module grease/git/enums/RepositoryProvider
 */

/**
 * Repository provider names.
 *
 * @enum {Lowercase<string>}
 */
enum RepositoryProvider {
  BITBUCKET = 'bitbucket',
  GITHUB = 'github',
  GITLAB = 'gitlab'
}

export default RepositoryProvider
