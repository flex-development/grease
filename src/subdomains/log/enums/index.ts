/**
 * @file Entry Point - Enums
 * @module grease/log/enums
 */

export { default as LogLevel } from './log-level'
export { default as UserLogLevel } from './log-level-user'
export { default as LogType } from './log-type'
