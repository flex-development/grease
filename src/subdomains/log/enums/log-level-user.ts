/**
 * @file Enums - UserLogLevel
 * @module grease/log/enums/UserLogLevel
 */

import type LogLevel from './log-level'

/**
 * User-friendly log levels.
 *
 * @enum {Lowercase<keyof typeof LogLevel>}
 */
enum UserLogLevel {
  DEBUG = 'debug',
  ERROR = 'error',
  FATAL = 'fatal',
  INFO = 'info',
  LOG = 'log',
  SILENT = 'silent',
  TRACE = 'trace',
  VERBOSE = 'verbose',
  WARN = 'warn'
}

export default UserLogLevel
