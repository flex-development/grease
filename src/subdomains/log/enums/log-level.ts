/**
 * @file Enums - LogLevel
 * @module grease/log/enums/LogLevel
 */

import type { Opaque } from '@flex-development/tutils'

/**
 * Log levels.
 *
 * @enum {number}
 */
enum LogLevel {
  DEBUG = 4,
  ERROR = 0,
  FATAL = -1,
  INFO = 3,
  LOG = 2,
  SILENT = <Opaque<number, '-INFINITY'>>Number.NEGATIVE_INFINITY,
  TRACE = 5,
  VERBOSE = <Opaque<number, 'INFINITY'>>Number.POSITIVE_INFINITY,
  WARN = 1
}

export default LogLevel
