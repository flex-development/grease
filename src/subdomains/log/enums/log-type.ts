/**
 * @file Enums - LogType
 * @module grease/log/enums/LogType
 */

/**
 * Log types.
 *
 * @enum {Lowercase<string>}
 */
enum LogType {
  DEBUG = 'debug',
  ERROR = 'error',
  FAIL = 'fail',
  FATAL = 'fatal',
  INFO = 'info',
  LOG = 'log',
  START = 'start',
  SUCCESS = 'success',
  TRACE = 'trace',
  VERBOSE = 'verbose',
  WARN = 'warn'
}

export default LogType
