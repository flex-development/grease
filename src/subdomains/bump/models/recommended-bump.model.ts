/**
 * @file Models - RecommendedBump
 * @module grease/bump/models/RecommendedBump
 */

import type { IRecommendedBump } from '#src/bump/interfaces'
import { ReleaseType } from '#src/enums'
import {
  ifelse,
  type Omit,
  type Simplify
} from '@flex-development/tutils'

/**
 * Recommended version bump data model.
 *
 * @see {@linkcode IRecommendedBump}
 *
 * @class
 * @implements {IRecommendedBump}
 */
class RecommendedBump implements IRecommendedBump {
  /**
   * Total number of breaking changes committed since last release.
   *
   * @public
   * @instance
   * @member {number} breaks
   */
  public breaks: number

  /**
   * Recommended release type.
   *
   * @see {@linkcode ReleaseType}
   *
   * @public
   * @instance
   * @member {Exclude<ReleaseType, 'prerelease'>} bump
   */
  public bump: Exclude<ReleaseType, 'prerelease'>

  /**
   * Number of commits considered.
   *
   * @public
   * @instance
   * @member {number} commits
   */
  public commits: number

  /**
   * Total number of features committed since last release.
   *
   * @public
   * @instance
   * @member {number} features
   */
  public features: number

  /**
   * Prerelease recommendation?
   *
   * @public
   * @instance
   * @member {boolean} unstable
   */
  public unstable: boolean

  /**
   * Create a new version bump recommendation.
   *
   * @see {@linkcode IRecommendedBump}
   *
   * @param {Omit<IRecommendedBump, 'bump'>} data - Version bump data
   */
  constructor(data: Omit<IRecommendedBump, 'bump'>) {
    this.breaks = data.breaks
    this.commits = data.commits
    this.features = data.features
    this.unstable = data.unstable

    switch (true) {
      case this.breaks > 0:
        this.bump = ReleaseType[ifelse(this.unstable, 'PREMAJOR', 'MAJOR')]
        break
      case this.features > 0:
        this.bump = ReleaseType[ifelse(this.unstable, 'PREMINOR', 'MINOR')]
        break
      default:
        this.bump = ReleaseType[ifelse(this.unstable, 'PREPATCH', 'PATCH')]
        break
    }
  }

  /**
   * Get a JSON-serializable recommendation.
   *
   * @public
   *
   * @return {Simplify<IRecommendedBump>} JSON-serializable recommendation
   */
  public toJSON(): Simplify<IRecommendedBump> {
    return {
      breaks: this.breaks,
      bump: this.bump,
      commits: this.commits,
      features: this.features,
      unstable: this.unstable
    }
  }
}

export default RecommendedBump
