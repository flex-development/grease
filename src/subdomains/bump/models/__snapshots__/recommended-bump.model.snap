// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:bump/models/RecommendedBump > #toJSON > should return json-serializable recommendation 1`] = `
{
  "breaks": 1,
  "bump": "premajor",
  "commits": 72,
  "features": 16,
  "unstable": true,
}
`;
